plugins/DQMProcessor.cpp

```diff
@@ -252,43 +272,62 @@ DQMProcessor::RequestMaker()
     // TLOG() << "Going to pop";
     try {
       m_source->pop(element, m_source_timeout);
-    } catch (const ers::Issue &excpt) {
+    } catch (const ers::Issue& excpt) {
       TLOG() << "DQM: Unable to pop from the data queue";
       continue;
     }
     ++m_data_count;
     ++m_total_data_count;
 
     TLOG_DEBUG(10) << "Data popped from the queue";
-    using runfunc_type = void (AnalysisModule::*)(std::unique_ptr<daqdataformats::TriggerRecord> record, std::unique_ptr<ChannelMap>& map, std::string kafka_address);
+    using runfunc_type = void (AnalysisModule::*)(std::unique_ptr<daqdataformats::TriggerRecord> record,
+                                                  std::unique_ptr<ChannelMap> & map,
+                                                  std::string kafka_address);
     runfunc_type memfunc = &AnalysisModule::run;
-    std::thread* current_thread = new std::thread(memfunc, std::ref(*algo), std::move(element), std::ref(m_map), m_kafka_address);
+    auto current_thread =
+      std::make_shared<std::thread>(memfunc, std::ref(*algo), std::move(element), std::ref(m_map), m_kafka_address);
 
     // Add a new entry for the current instance
     TLOG() << "Starting to run \"" << analysis_instance.name << "\"";
     map[std::chrono::system_clock::now() +
         std::chrono::milliseconds(static_cast<int>(analysis_instance.between_time) * 1000)] = {
-      algo, analysis_instance.between_time, analysis_instance.default_unavailable_time,
-      analysis_instance.number_of_frames, current_thread, analysis_instance.name
+      algo,
+      analysis_instance.between_time,
+      analysis_instance.default_unavailable_time,
+      analysis_instance.number_of_frames,
+      current_thread,
+      analysis_instance.name
     };
 
-    // Delete the entry we just used and find the next one
-    map.erase(task);
-
     // Delete thread
     if (previous_thread != nullptr) {
       if (previous_thread->joinable()) {
         previous_thread->join();
-        delete previous_thread; // TODO: rsipos -> Why is this a raw pointer on the thread? Move to unique_ptr.
       } else {
-        TLOG() << "Thread not joinable";
-        // Should not be happening
+        throw ProcessorError(ERS_HERE, "Thread not joinable");
       }
     }
+
+    // Delete the entry we just used and find the next one
+
+    // Note that since previous_thread refers to the thread
+    // corresponding to task, it's safe to now remove the
+    // AnalysisInstance whose thread it refers to
+
+    map.erase(task);
+
+
   }
+
+  for (auto& [time, analysis_instance] : map) {
+    if (analysis_instance.running_thread && analysis_instance.running_thread->joinable()) {
+      analysis_instance.running_thread->join();
+    }
+  }
+
```

Comment:
If there is a task that takes longer than the timeout of the stop command in nanorc, it will crash on stop